### PR TITLE
CAP-290 harmonise metadata with edr pt2

### PR DESCRIPTION
**Pull request by crbunney**
Created:
* `Duration` class to represent ISO8601 Durations (as described in https://en.wikipedia.org/wiki/ISO_8601#Durations)
  * Allows both the parsing of valid strings and the serialisation of objects to valid strings
  * Supports addition & subtraction between other `Duration`s, `datetime.datetime`, and `datetime.timedelta`
  * Supports conversion to/from `datetime.timedelta` and `dateutil`'s `relativedelta`
* `DateTimeInterval` to represent ISO8601 Time Intervals (as described in https://en.wikipedia.org/wiki/ISO_8601#Time_intervals)
  * Supports the inference of properties from available information (e.g. if start datetime and duration are provided, when `DateTimeInterval.end` is called, the end datetime is inferred and returned; if start and end datetimes are provided, infer the duration, etc)
  * Provides `upper_bound`, `lower_bound`, and `total_duration` properties which ensure consistent handling between different ways of creating equivalent DateTimeIntervals and are also take recurrences (both finite and infinite) into account
* `TemporalExtent` to store the time information for a dataset
  * Allows a combination of specific datetimes and `DateTimeIntervals`
  * `TemporalExtent.bounds()`calculates and returns the beginning and end of the time period covered by the extent

Updated `Metadata` (de)serialisation to persist the new information introduced by the above changes.

### CODE REVIEW

**Review by DPeterK — COMMENTED** [review]
First off, good work implementing functionality that handles all the vagaries of the full ISO8601 date/time/durations specification! It is evidently full-featured and not without challenge to produce.

I do have a significant question about the specifics of the implementation - particularly the number of private static methods that return new instances of `Duration` rather than updating a current instance. To my eyes at least it's making for duplicated logic and a code flow that's quite hard to follow. I _think_ there's two main reasons you're using this pattern:
* (primarily) for error-checking and pre-processing ahead of constructing a Duration instance
* (occasionally) to utilise encapsulated logic in the class without also needing an instance of the class.

I might well be missing something extra here that's required the design you've followed, but I wonder if an alternate design might be able to remove these static methods:
* you could override `__new__` to encapsulate necessary pre-init logic and error-checking, while retaining the more typical `from_<other>` constructor static methods
* you could move the ISO8601 parsing into a helper function that's referenced in methods in both `Duration` and `DateTimeInterval`.

**Comment by DPeterK** [comment]
I'd still like to see the final version of this code added/available to the EDR Server (perhaps in its own library?) as the server needs to make use of similar functionality - for parsing datetime strings and generating lists of datetime strings from intervals.

While we're on the subject of separate libraries, did you confirm that neither [Arrow](https://arrow.readthedocs.io/en/latest/) nor dateutil (particularly [isoparse](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse)) do some or all of what the code here does? I do remember some conversation about prior art, but don't remember if these libraries were included in that discussion.

File: src/clean_air/models.py [hunk 1]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+
+    # Constants used during various field conversions
+    _MONTHS_IN_YEAR = 12
+    _DAYS_IN_WEEK = 7
+    _MICROSECONDS_IN_SECOND = 1000000
+    _SECONDS_IN_MINUTE = 60
+    _SECONDS_IN_HOUR = 60 * _SECONDS_IN_MINUTE
+    _SECONDS_IN_DAY = 24 * _SECONDS_IN_HOUR
+    _SECONDS_IN_WEEK = 7 * _SECONDS_IN_DAY
+    _SECONDS_IN_YEAR = 365 * _SECONDS_IN_DAY  # Common year, rather than Gregorian year (and definitely not Julian!)
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:89]: Yes, Rachel also mentioned to me that some model runs are based on a standardised 30-day long month, which also triggered a headache.
The 365 day year is at least consistent with the `dateutil` library (which I made use of)

File: src/clean_air/models.py [hunk 2]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+    _PARSER = re.compile("".join([
+        "^P",
+        _FLOAT_REGEX_FMT.format("years", "Y"),
+        _FLOAT_REGEX_FMT.format("months", "M"),
+        _FLOAT_REGEX_FMT.format("days", "D"),
+        "(?:T",  # T is required if any of the following fields are present
+        _FLOAT_REGEX_FMT.format("hours", "H"),
+        _FLOAT_REGEX_FMT.format("minutes", "M"),
+        _FLOAT_REGEX_FMT.format("seconds", "S"),
+        ")?",
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:105]: yea, tell me about it...

File: src/clean_air/models.py [hunk 3]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+        """
+
+        # Allowing Nones and converting them here makes it easier to call this method
+        years = years if years is not None else 0
+        months = months if months is not None else 0
+        days = days if days is not None else 0
+        hours = hours if hours is not None else 0
+        minutes = minutes if minutes is not None else 0
+        seconds = seconds if seconds is not None else 0
+        weeks = weeks if weeks is not None else 0
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:215]: This logic is reused during the in the normalise function, which is used in some contexts before we can create an actual `Duration` object, hence why it's static and why we're duplicating some of the functionality from the properties.

File: src/clean_air/models.py [hunk 4]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+            # without a value
+            kwargs = {k: v for k, v in kwargs.items() if v}
+            # relative deltas can contain negative fields, so we need to normalise the result
+            return Duration._normalise(**kwargs)
+        elif isinstance(obj, timedelta):
+            return Duration._normalise(seconds=obj.total_seconds())
+        else:
+            raise NotImplementedError(f"{type(obj)} is not supported")
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:276]: Essentially the rationale for making them static is that they don't access any instance attributes.
The reason for that is that they're used in methods like `__add__` and `__sub__` when we're doing arithmetic with things like `datetime`, `timedelta`, and `relativedelta`. I.e. we don't actually have a `Duration` object but still need to perform the calculation. It also means I don't have to create multiple `Duration` objects, and in particular there are cases where I can't create a Duration until after I've normalised the values.

e.g. relativedelta supports negatives where Duration doesn't, which causes problems in cases like this:
`Duration(years=1) + relativedelta(days=-1)`. If we don't normalise the fields, we end up trying to create `Duration(years=1, days=-1)` which throws a `ValueError`. Normalisation converts that to `Duration(days=364)`.

File: src/clean_air/models.py [hunk 5]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+            raise ValueError("end cannot be specified on its own. Please provide either and end date or a duration")
+        if recurrences is not None and recurrences < -1:
+            raise ValueError("recurrences cannot be less than -1")
+
+        self._start = start
+        self._end = end
+        self._duration = duration
+        self._recurrences = recurrences
+
+    @cached_property
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:663]: Yes, although I decided to do some premature optimisation given I felt there was enough stuff going on for what is meant to be an unchanging value (Duration objects are designed to be immutable)

File: src/clean_air/models.py [hunk 6]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+    def _id(self):
+        """
+        This is a helper method that standardises how we compare Duration instances.
+        It's used in functions like __hash__, __eq__, __lte__, etc
+
+        Cached because it requires various calculations and has the potential to be called a lot.
+        (We can cache it because Duration is designed to be immutable)
+        """
+        return Duration._get_months_seconds(**Duration._to_dict(self))
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:185]: I like to be explicit when I'm accessing static methods, otherwise, there's not any indication to the reader that it is a static method unless they look up the definition

File: src/clean_air/models.py [hunk 7]
```diff
@@ -27,21 +34,879 @@ class ContactDetails:
...
+        _FLOAT_REGEX_FMT.format("minutes", "M"),
+        _FLOAT_REGEX_FMT.format("seconds", "S"),
+        ")?",
+    ]), re.ASCII)
+    # Regex for parsing and extracting the special case of weeks, which can't be combined with any of the other fields
+    _WEEK_PARSER = re.compile(r"^P(?P<weeks>\d+(?:\.\d+)?)W$", re.ASCII)
+
+    def __init__(
+            self,
+            years: Optional[float] = None,
```
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:112]: `Optional[float]` actually implies `Union[float, None]`

**Comment by crbunney** [comment]
> I'd still like to see the final version of this code added/available to the EDR Server (perhaps in its own library?) as the server needs to make use of similar functionality - for parsing datetime strings and generating lists of datetime strings from intervals.

Yep, I agree. My plan was to get it reviewed and integrated here, and then look at getting it into the edr-server (and subsequently removing it from here)
 
> While we're on the subject of separate libraries, did you confirm that neither [Arrow](https://arrow.readthedocs.io/en/latest/) nor dateutil (particularly [isoparse](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse)) do some or all of what the code here does? I do remember some conversation about prior art, but don't remember if these libraries were included in that discussion.

Yea, those ones only do the date, time, and timestamp parts of the 8601 spec. I do actually use [isoparse](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse) for the date parsing part of `DateTimeInterval`, but there were no good candidates for the `Duration` logic, or `DateTimeInterval`. It seems these bits are too niche

File: src/clean_air/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DPeterK) [reply src/clean_air/models.py:105]: I really don't like all this WKT stuff (which this ISO format string is effectively an example of) - it seems exessively hard to parse on a computer while a person can pull all the information out quickly and easily. I spent quite a while trying in vain to find a system to parse the CRS / VRS / TRS format strings used in the EDR spec (I got as far as abstract syntax trees before deciding that really they just recreated the problem in a whole load of Python objects rather than a single string) before deciding I didn't need them anyway. I just want a system that pulls all the important stuff out of the WKT and stores in in JSON or something...

File: src/clean_air/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DPeterK) [reply src/clean_air/models.py:276]: I guess what I was thinking was that if `__new__` was used to encapsulate the pre-processing necessary then we'd always have valid Duration instances and fewer intermediates - and, in my mind at least, it would feel more Pythonic. But that's an implementation detail, and `@staticmethod` is pretty Pythonic!

**Comment by DPeterK** [comment]
> My plan was to get it reviewed and integrated here
 Good plan.

> It seems these bits are too niche
which seems a bit of a cop-out given it's part of the spec!

**Comment by DPeterK** [comment]
Here's another ISO8601 parser that I was pointed to this morning - it supports intervals as well: https://pypi.org/project/aniso8601/

File: src/clean_air/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (crbunney) [reply src/clean_air/models.py:276]: Yea, I did have that thought, but I wanted to avoid forcing all `Duration`s to be normalised, as then the parsing function would become lossy/would no longer represent the duration string as given. I can imagine cases where someone might want to keep `P36H` as `Duration(hours=36)` rather than be normalised to `Duration(days=1, hours=12)`. E.g. when dealing with timesteps relating to model runs they tend to be all expressed in un-normalised hours.

So it's not that I didn't want to do what you suggest, but it seemed incompatible with some of the (admittedly self-imposed) constraints on the behaviour!
